### PR TITLE
Azure / e2e Cody suite

### PR DESCRIPTION
**Pull request by cbart**
Part of https://github.com/sourcegraph/sourcegraph/issues/56182

Here's how to get this running:

1. Azure keys
   * https://portal.azure.com > Azure OpenAI > sourcegraph-test-oai > Keys and Endpoint
   * Copy the key and `export AZURE_API_KEY="<paste the key here>"`
1. `export SRC_ENDPOINT=https://sourcegraph.sourcegraph.com`
1. `export SRC_ACCESS_TOKEN="<s2 access token goes here>"`
1. Run like so: `pnpm run start --label="Beam search" --output="$HOME/test.results" --provider=azure`
   * single beam search test runs faster than total # test cases
   * output file actually allows you to drill into the LLM interaction details

## Test plan

Manually verified:

### success scenario with Azure OpenAI and

```
➜  e2e git:(cb/s/56182-azure) ✗ pnpm run start --label="Beam search" --output="$HOME/azure.results" --provider=azure

> @sourcegraph/cody-e2e@0.0.1 start /Users/cbart/cody/e2e
> pnpm run --silent build && node dist/e2e "--label=Beam search" "--output=/Users/cbart/azure.results" "--provider=azure"

Run 1 of 1

Testing (1/1): Beam search (github.com/huggingface/transformers)
Incorrect answers (LLM judge): 0/2 (0%)
Incorrect or partial answers (LLM judge): 1/2 (50%)
Missing facts: 5/9 (56%)
Hallucinated entities: 9/9 (100%)

Run 1 of 1 summary:
Incorrect answers (LLM judge): 0/2 (0%)
Incorrect or partial answers (LLM judge): 1/2 (50%)
Missing facts: 5/9 (56%)
Hallucinated entities: 9/9 (100%)
```

### fast fail when not enough env vars provided

```
➜  e2e git:(cb/s/56182-azure) unset AZURE_DEPLOYMENT_ID
➜  e2e git:(cb/s/56182-azure) ✗ pnpm run start --label="Beam search" --output="$HOME/azure.results" --provider=azure

> @sourcegraph/cody-e2e@0.0.1 start /Users/cbart/cody/e2e
> pnpm run --silent build && node dist/e2e "--label=Beam search" "--output=/Users/cbart/azure.results" "--provider=azure"

Missing required environment variables: AZURE_DEPLOYMENT_ID
1. https://portal.azure.com > Azure OpenAI > sourcegraph-test-oai > Keys and Endpoint
2. Go to Model Deployments > Manage Deployments. Find the deployment name and:
   export AZURE_DEPLOYMENT_ID="<paste deployment name here>"
 ELIFECYCLE  Command failed with exit code 1.
```

### CODE REVIEW

**Comment by cbart** [comment]
We're getting different judgements for Anthropic and Azure:

```
➜  e2e git:(cb/s/56182-azure) ✗ pnpm run start --label="Beam search" --output="$HOME/azure.results" --provider=azure

> @sourcegraph/cody-e2e@0.0.1 start /Users/cbart/cody/e2e
> pnpm run --silent build && node dist/e2e "--label=Beam search" "--output=/Users/cbart/azure.results" "--provider=azure"

Run 1 of 1

Testing (1/1): Beam search (github.com/huggingface/transformers)
Incorrect answers (LLM judge): 0/2 (0%)
Incorrect or partial answers (LLM judge): 1/2 (50%)
Missing facts: 5/9 (56%)
Hallucinated entities: 9/9 (100%)

Run 1 of 1 summary:
Incorrect answers (LLM judge): 0/2 (0%)
Incorrect or partial answers (LLM judge): 1/2 (50%)
Missing facts: 5/9 (56%)
Hallucinated entities: 9/9 (100%)

➜  e2e git:(cb/s/56182-azure) ✗ pnpm run start --label="Beam search" --output="$HOME/anthropic.results" --provider=anthropic

> @sourcegraph/cody-e2e@0.0.1 start /Users/cbart/cody/e2e
> pnpm run --silent build && node dist/e2e "--label=Beam search" "--output=/Users/cbart/anthropic.results" "--provider=anthropic"

Run 1 of 1

Testing (1/1): Beam search (github.com/huggingface/transformers)
Incorrect answers (LLM judge): 0/2 (0%)
Incorrect or partial answers (LLM judge): 0/2 (0%)
Missing facts: 4/9 (44%)
Hallucinated entities: 9/9 (100%)

Run 1 of 1 summary:
Incorrect answers (LLM judge): 0/2 (0%)
Incorrect or partial answers (LLM judge): 0/2 (0%)
Missing facts: 4/9 (44%)
Hallucinated entities: 9/9 (100%)
```

File: e2e/src/llm-judge.ts [hunk 1]
```diff
@@ -43,18 +47,44 @@ export async function llmJudge(
...
         answerMatchesSummary: doesAnswerMatchSummary(completion.completion),
         answerMatchesSummaryJudgement: completion.completion,
     }
 }
 
+async function azureJudge(instructions: string): Promise<LLMJudgement> {
+    // 1. https://portal.azure.com > Azure OpenAI > sourcegraph-test-oai > Keys and Endpoint
+    // 2. Copy the key and export AZURE_API_KEY="<paste the key here>"
+    const azureApiKey = process.env.AZURE_API_KEY as string
+    const endpoint = 'https://sourcegraph-test-oai.openai.azure.com/'
```
Comment (cbart) [inline e2e/src/llm-judge.ts:74]: Probably need to move the endpoint and the deploymentId to ENV. @novoselrok well this is a bit prototype-y but like, could we land it and iterate?

Comment (novoselrok) [reply e2e/src/llm-judge.ts:74]: Moving the `endpoint` to an env variable would be great.

File: e2e/src/llm-judge.ts [hunk 2]
```diff
@@ -43,18 +47,44 @@ export async function llmJudge(
...
+
+    const client = new OpenAIClient(endpoint, new AzureKeyCredential(azureApiKey))
+    const deploymentId = 'gpt-35-turbo-test'
+    const result = await client.getCompletions(deploymentId, [instructions], { maxTokens: 128 })
```
Comment (novoselrok) [inline e2e/src/llm-judge.ts:78]: We sample `300` tokens from Anthropic (line 61). We can extract that to a separate const and use it here as well.

**Review by novoselrok — APPROVED** [review]

**Comment by novoselrok** [comment]
Before merging, can you also stamp https://github.com/sourcegraph/cody/pull/705 so I can merge it and you can check for any conflicts?

**Comment by cbart** [comment]
> Before merging, can you also stamp #705 so I can merge it and then check for any conflicts?

Will do!

**Comment by cbart** [comment]
I've added some fail-fast checks for environment variables to mitigate the scenario where one forgets say AZURE_DEPLOYMENT_ID and runs the whole test suite just to learn that they cannot validate.

**Comment by cbart** [comment]
All comments addressed. This is ready to land.

**Pull request merged**